ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarRecoveringCustodyImpl.java

```diff
@@ -0,0 +1,290 @@
+  }
+
+  private synchronized void createOrUpdateRecoveryTaskForBlock(final BeaconBlock block) {
+    if (recoveryTasks.containsKey(block.getSlotAndBlockRoot())) {
+      final RecoveryTask existing = recoveryTasks.get(block.getSlotAndBlockRoot());
+      if (existing.block().get() == null) {
+        existing.block().set(block);
+        maybeStartRecovery(existing);
+      }
+    } else {
+      recoveryTasks.put(
+          block.getSlotAndBlockRoot(),
+          new RecoveryTask(
+              new AtomicReference<>(block),
+              new HashSet<>(),
+              new AtomicBoolean(false),
+              new AtomicBoolean(false)));
+    }
+  }
+
+  private synchronized void maybeStartRecovery(final RecoveryTask task) {
+    if (readyToBeRecovered(task)) {
+      task.recoveryStarted().set(true);
+      if (task.existing().size() != columnCount) {
+        asyncRunner.runAsync(() -> prepareAndInitiateRecovery(task)).ifExceptionGetsHereRaiseABug();
+      }
+    }
+  }
+
+  private boolean readyToBeRecovered(final RecoveryTask task) {
+    if (task.recoveryStarted().get()) {
+      // already started
+      return false;
+    }
+
+    if (!task.timeouted().get()) {
+      return false;
+    }
+
+    if (task.existing().size() < recoverColumnCount) {
+      // not enough columns collected
+      return false;
+    }
+
+    if (task.block().get() == null) {
+      return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public void subscribeToValidDataColumnSidecars(
+      final DataColumnSidecarManager.ValidDataColumnSidecarsListener sidecarsListener) {
+    validDataColumnSidecarsSubscribers.subscribe(sidecarsListener);
+  }
+
+  private boolean isActiveSuperNode(final UInt64 slot) {
+    return isSuperNode
+        && spec.atSlot(slot).getMilestone().isGreaterThanOrEqualTo(SpecMilestone.FULU);
+  }
+
+  private record RecoveryTask(
+      AtomicReference<BeaconBlock> block,
+      Set<DataColumnSlotAndIdentifier> existing,
+      AtomicBoolean recoveryStarted,
+      AtomicBoolean timeouted) {}
+
+  private void prepareAndInitiateRecovery(final RecoveryTask task) {
+    final SafeFuture<List<DataColumnSidecar>> list =
+        AsyncStream.create(task.existing().stream())
+            .mapAsync(delegate::getCustodyDataColumnSidecar)
+            .map(Optional::get)
+            .toList();
+    initiateRecovery(task.block().get(), list);
+  }
+
+  private void initiateRecovery(
+      final BeaconBlock block, final SafeFuture<List<DataColumnSidecar>> list) {
+    LOG.info("Starting data columns sidecars recovery for block: {}", block.getSlotAndBlockRoot());
+
+    list.thenPeek(
+            sidecars -> {
+              LOG.debug(
+                  "Recovery for block: {}. DatacolumnSidecars found: {}",
+                  block.getSlotAndBlockRoot(),
+                  sidecars.size());
+              final Map<UInt64, DataColumnSidecar> existingSidecarsByColIdx = new HashMap<>();
```

Comment:
nit: We need just a set of indexes here:  
```java
              Set<UInt64> existingSidecarIndexes =
                  sidecars.stream().map(DataColumnSidecar::getIndex).collect(Collectors.toSet());
```

Reply:
Nice catch!